libsolidity/analysis/ControlFlowGraph.cpp

```diff
@@ -0,0 +1,552 @@
+			if (node != mergeDestination)
+				connect(node, mergeDestination);
+		m_currentNode = mergeDestination;
+	}
+
+	CFGNode* newLabel()
+	{
+		return m_cfg.newNode();
+	}
+	CFGNode* createLabelHere()
+	{
+		auto label = m_cfg.newNode();
+		connect(m_currentNode, label);
+		m_currentNode = label;
+		return label;
+	}
+	void placeAndConnectLabel(CFGNode *_node)
+	{
+		connect(m_currentNode, _node);
+		m_currentNode = _node;
+	}
+
+	CFG& m_cfg;
+	CFGNode* m_currentNode = nullptr;
+
+	/// The current jump destination of break Statements.
+	CFGNode* m_breakJump = nullptr;
+	/// The current jump destination of continue Statements.
+	CFGNode* m_continueJump = nullptr;
+
+	/// Helper class that replaces the break and continue jump destinations for the
+	/// current scope and restores the originals at the end of the scope.
+	class BreakContinueScope
+	{
+	public:
+		BreakContinueScope(ControlFlowParser& _parser, CFGNode* _breakJump, CFGNode* _continueJump):
+			m_parser(_parser), m_origBreakJump(_parser.m_breakJump), m_origContinueJump(_parser.m_continueJump)
+		{
+			m_parser.m_breakJump = _breakJump;
+			m_parser.m_continueJump = _continueJump;
+		}
+		~BreakContinueScope()
+		{
+			m_parser.m_breakJump = m_origBreakJump;
+			m_parser.m_continueJump = m_origContinueJump;
+		}
+	private:
+		ControlFlowParser& m_parser;
+		CFGNode* m_origBreakJump;
+		CFGNode* m_origContinueJump;
+	};
+};
+
+}
+}
+
+bool ControlFlowParser::visit(BinaryOperation const& _operation)
+{
+	solAssert(!!m_currentNode, "");
+
+	switch(_operation.getOperator())
+	{
+		case Token::Or:
+		case Token::And:
+		{
+			appendControlFlow(_operation.leftExpression());
+
+			auto nodes = splitFlow<2>();
+			nodes[0] = createFlow(nodes[0], _operation.rightExpression());
+			mergeFlow(nodes, nodes[1]);
+
+			return false;
+		}
+		default:
+			break;
+	}
+	return ASTConstVisitor::visit(_operation);
+}
+
+bool ControlFlowParser::visit(Conditional const& _conditional)
+{
+	solAssert(!!m_currentNode, "");
+
+	_conditional.condition().accept(*this);
+
+	auto nodes = splitFlow<2>();
+
+	nodes[0] = createFlow(nodes[0], _conditional.trueExpression());
+	nodes[1] = createFlow(nodes[1], _conditional.falseExpression());
+
+	mergeFlow(nodes);
+
+	return false;
+}
+
+bool ControlFlowParser::visit(IfStatement const& _ifStatement)
+{
+	solAssert(!!m_currentNode, "");
+
+	_ifStatement.condition().accept(*this);
+
+	if (_ifStatement.falseStatement())
+	{
+		auto nodes = splitFlow<2>();
+		nodes[0] = createFlow(nodes[0], _ifStatement.trueStatement());
+		nodes[1] = createFlow(nodes[1], *_ifStatement.falseStatement());
+		mergeFlow(nodes);
+	}
+	else
+	{
+		auto nodes = splitFlow<2>();
+		nodes[0] = createFlow(nodes[0], _ifStatement.trueStatement());
+		mergeFlow(nodes, nodes[1]);
+	}
+
+	return false;
+}
+
+bool ControlFlowParser::visit(ForStatement const& _forStatement)
+{
+	solAssert(!!m_currentNode, "");
+
+	if (auto initializationExpression = _forStatement.initializationExpression())
```

Comment:
If we ever change the return type here, this might perform a copy without us noticing. I think it's fair to just call the function twice...

Reply:
If the return type is changed to something that might perform a copy, the pointer dereference in the next line should fail, though. I can change to ``auto const*``, unless that's still too much ``auto`` :)..